File: c/data_structures/tree/traversal/inorder.c
#include "../binaryTree.c"

// Perform a inorder traversal of the BST where the result is: { 25, 50, 75, 100, 110, 125, 150, 175 }
//
//
//              __ 100 __
//             /         \
//         50              150
//       /    \          /     \
//     25      75      125     175
//                   /
//                 110
//
//

int n = 0;
void inorder(int tree[], struct node *root) {
    if (!root)
        return;

    if (root->left)
        inorder(tree, root->left);

    *(tree + n) = root->value;
    n++;

    if (root->right)
        inorder(tree, root->right);

    return;
}

int main(void) {
    int tree[8];

    struct node *root = makeNode(100);
    struct node *a = makeNode(50);
    struct node *b = makeNode(150);
    struct node *c = makeNode(25);
    struct node *d = makeNode(75);
    struct node *e = makeNode(125);
    struct node *f = makeNode(175);
    struct node *g = makeNode(110);

    root->left = a;
    root->right = b;

    a->left = c;
    a->right = d;

    b->left = e;
    b->right = f;

    e->left = g;

    inorder(tree, root);

    printf("{ ");
    for (n = 0; n < 8; n++, *(tree + n)) {
        printf("%d ", tree[n]);
    }
    printf(" }");

    return 0;
}

File: python/tree/traversal/inorder.py
import ipdb

from Tree import TreeNode


# Recursive
# Left -> Node -> Right
#def inorder(root, visited=[]):
#    if not root:
#        return
#
#    inorder(root.left, visited)
#    visited.append(root.value)
#    inorder(root.right, visited)
#
#    return visited


# Iterative
# Left -> Node -> Right
#def inorder(root):
#    # Unlike other iterative traversals, we don't start with the root
#    # already in the stack.
#    stack = []
#    current = root
#    visited = []
#
#    while current or stack:
#        # Traverse as far left as we can.
#        while current:
#            stack.append(current)
#            current = current.left
#
#        # Pop the leftmost node.
#        current = stack.pop()
#        visited.append(current.value)
#
#        # Assume there's a right node.  If not, the next left node will
#        # be popped, and then we'll again assume there's a right node.
#        current = current.right
#    return visited


# Morris traversal.
def inorder(root):
    current = root
    res = []

    while current:
        if not current.left:
            res.append(current.value)
            current = current.right
        else:
            pre = current.left
            # Find the inorder predecessor of current.
            # This will be the right-most node of the left subtree.
            while pre.right:
                pre = pre.right
            pre.right = current

            # This is destroying the original tree which is no bueno!
            tmp = current
            current = current.left
            tmp.left = None

    return res


#nodes = [1, 2, 3, 4, 5, 6]
root = TreeNode(1)
root.left = TreeNode(2)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)
root.right = TreeNode(3)
root.right.right = TreeNode(6)

#               1
#             /   \
#            /     \
#           2       3
#          / \       \
#         /   \       \
#        4     5       6
#
#  inorder [4, 2, 5, 1, 3, 6]

print(inorder(root))
